pkg/ensign/api/v1beta1/deduplication.go

```diff
@@ -0,0 +1,194 @@
+	}
+
+	// Write the type to the hash
+	var etype *Type
+	if event.Type != nil && !event.Type.IsZero() {
+		etype = event.Type
+	} else {
+		etype = UnspecifiedType
+	}
+
+	if _, err = hash.Write([]byte(etype.Name)); err != nil {
+		return nil, fmt.Errorf("could not write type name to hash: %w", err)
+	}
+
+	if err = binary.Write(hash, binary.LittleEndian, etype.MajorVersion); err != nil {
+		return nil, fmt.Errorf("could not hash type major version: %w", err)
+	}
+
+	if err = binary.Write(hash, binary.LittleEndian, etype.MinorVersion); err != nil {
+		return nil, fmt.Errorf("could not hash type minor version: %w", err)
+	}
+
+	if err = binary.Write(hash, binary.LittleEndian, etype.PatchVersion); err != nil {
+		return nil, fmt.Errorf("could not hash type patch version: %w", err)
+	}
+
+	return hash.Sum(nil), nil
+}
+
+// Datagram hashing is used to detect duplicates in data only, ignoring metadata,
+// mimetype, and type as in strict hashing. This method returns a murmur3 hash of the
+// data field of the event only.
+func (w *EventWrapper) HashDatagram() (_ []byte, err error) {
+	var event *Event
+	if event, err = w.Unwrap(); err != nil {
+		return nil, err
+	}
+
+	hash := murmur3.New128()
+	if _, err = hash.Write(event.Data); err != nil {
+		return nil, err
+	}
+
+	return hash.Sum(nil), nil
+}
+
+// Key grouped hashing returns the murmur3 hash of the data of the event prefixed with
+// the metadata values of the the specified keys. E.g. if the data is foobar and the
+// hash is grouped by the key month - then for two events with month jan and month feb
+// will have different hashes: murmur3(janfoobar) and murmur3(febfoobar).
+//
+// NOTE: this method does not take into account mimetype or type but in the future we
+// may have "reserved keys" to factor in these elements to the hash.
+func (w *EventWrapper) HashKeyGrouped(keys []string) (_ []byte, err error) {
+	if len(keys) == 0 {
+		return nil, ErrNoKeys
+	}
+
+	var event *Event
+	if event, err = w.Unwrap(); err != nil {
+		return nil, err
+	}
+
+	hash := murmur3.New128()
+	for _, key := range keys {
+		val := event.Metadata[key]
+		if _, err = hash.Write([]byte(key + val)); err != nil {
+			return nil, err
+		}
+	}
+
+	if _, err = hash.Write(event.Data); err != nil {
+		return nil, err
+	}
+
+	return hash.Sum(nil), nil
+}
+
+// Unique key hashes determine duplicates not from the event data but from the keys
+// specified in the metadata (useful for creating lookup indexes). The hash is the
+// murmur3 hash of the concatenated key values for the specified keys.
+func (w *EventWrapper) HashUniqueKey(keys []string) (_ []byte, err error) {
+	if len(keys) == 0 {
+		return nil, ErrNoKeys
+	}
+
+	var event *Event
+	if event, err = w.Unwrap(); err != nil {
+		return nil, err
+	}
+
+	hash := murmur3.New128()
+	for _, key := range keys {
+		val := event.Metadata[key]
+		if _, err = hash.Write([]byte(key + val)); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
See above comment about hashing key/value, policy order, and whether or not the concatenation is necessary.